### ISSUE TITLE
`optional` is too constrained.

### ISSUE THREAD

**Issue by HuwCampbell** [purescript/purescript-maybe#45]
I don't think optional needs the `Plus` part of `Alternative`.

```haskell
optional :: forall f a. Alt f => Applicative f => f a -> f (Maybe a)
optional a = map Just a <|> pure Nothing
```

I have had to locally fork this for an experiment I'm doing.

**Comment by HuwCampbell** [comment]
The reason is that `Plus` requires that an empty element exists, but that's not always true.

**Comment by hdgarrood** [comment]
I think this makes sense. For a concrete use case, I can see the less constrained version being useful for a parsing library, where you might want to provide an `Alt` instance for trying one parser and then trying some other parser if the first one doesn't succeed, but where you don't want to provide a `Plus` instance, because that requires you to introduce the concept of an "empty" (i.e. useless) error message.

Haskell's `semigroupoids` library provides the less constrained version of `optional` too: https://hackage.haskell.org/package/semigroupoids-5.3.4/docs/Data-Functor-Alt.html#v:optional

**Comment by hdgarrood** [comment]
Ha, I just read your optparse PR and realised that's pretty much exactly your reasoning too! I came across this same issue independently in https://github.com/hdgarrood/aeson-better-errors/issues/12.

**Comment by hdgarrood** [comment]
Resolved by #47. Thanks"
